[MRG] ENH: Optional positivity constraints on the dictionary and sparse code

jakirkham opened the pull request:
Allows for a positivity constraint to be set during dictionary learning. This should be very similar to one provided for by Marial, et al. in SPAMS. However, I may need guidance in making sure this is working correctly in all cases.

**Todo:**
- [x] Positivity constraint for dictionary.
  - [x] Provide support for positivity constraint on dictionary.
  - [x] API documentation for using positivity constraint.
  - [x] Tests demonstrating the positivity constraint works correctly.
- [X] Positivity constraint for sparse code.
  - [X] Provide support for positivity constraint on sparse code.
  - [X] API documentation for using positivity constraint.
  - [X] Raise exception when positivity constraint is set incorrectly (when using orthogonal matching pursuit).
  - [X] Tests demonstrating the positivity constraint works correctly.
- [x] Add an example of using this functionality.

GaelVaroquaux commented:
In sparse_encode, you cannot ignore the positive argument for algorithms that do not support it.  You need either to support it (for threshold it's just a question of putting the negative values to zero) or to raise an error. 

In addition, you need a test for all the features that you have added,  here testing positive support for all algorithms.

jakirkham commented:
Added positivity support for the threshold case and added a `ValueError` for the orthogonal matching pursuit case.

Added tests that verify the positivity constraint is met for the sparse code in a variety of cases. Also, verifies that the `ValueError` for the orthogonal matching pursuit case is raised appropriately.

It would be nice to extend this to the dictionary, as well. This is something that Marial, et al. do in SPAMS. I know this will come down to some sort of thresholding on the dictionary, but am a little unclear as to where. If you have any thoughts, please share.

jakirkham commented:
So, I have added something that I think will correctly constrain the dictionary to positive values. It does appear to work. However, it may not be the best way or could have errors so feedback is definitely welcome here.

jakirkham commented:
Thus far, using this function on more complex datasets seems to yield the right results.

arthurmensch commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -304,7 +321,7 @@ def sparse_encode(X, dictionary, gram=None, cov=None, algorithm='lasso_lars',
 
 def _update_dict(dictionary, Y, code, verbose=False, return_r2=False,
-                 random_state=None):
+                 random_state=None, dict_positive=False):
```
`positive=False`

jakirkham replied:
I had thought about this, but was concerned it might be unclear. If you would prefer it, I can make the change.

jakirkham replied:
This has been changed.

arthurmensch commented:
We need an example showing the interest of enforcing positivity to the dictionary / code. It would be great to compare to NMF estimators. I guess modifying faces decomposition examples is the easiest way, though it would be interesting to have a different usecase.

jakirkham commented:
So, I am using this in image processing of calcium image data from the mouse brain primarily (various regions). While there are many cases of using NMF in this area (and we tried ourselves), we found dictionary learning with positivity constraints was more performant. Even on artificial data (used for testing), I will get the wrong answer without this positivity constraint on the dictionary.

The reason I added this is it was a feature provided in SPAMS, which Marial, et al. provided. However, this package has gone without a release for a year and a half. I have tried patching it to fix various bugs that have crept up (NumPy support, better linking to the BLAS, etc.), but this solution is untenable. Especially as I am now looking to have Python 3 support and a new version of NumPy is on the horizon. It seems the original authors are likely exploring new interesting problems, which makes sense. Unfortunately, this is a problem from the maintenance side. I suppose one could argue about licensing too, but this seems secondary to having a working dependency stack.

arthurmensch commented:
Can you show that it outperforms NMF on faces for instance ?

jakirkham commented:
Ok, sure, I'll try to come up with an example. It might not be today, but I will at least try tomorrow or this weekend.

jakirkham commented:
For an example, take a look at this ( http://nbviewer.jupyter.org/gist/jakirkham/d6e3921bad5a1d263990 ).

jakirkham commented:
Any thoughts on the example, @arthurmensch? I would really like to find a way forward here, but am unclear as to what it is at this point.

jakirkham commented:
Any more thoughts on this?

amueller commented:
Sorry for our slow replies @jakirkham. I think we don't have maintainers that use these methods a lot unfortunately, and we are swamped with many PRs. We need to figure something out though :-/

jnothman commented:
Ping @vene

vene commented:
(Caveat: I don't really have any experience with practical application of dictionary learning whatsoever...)

Maybe just show the last slice of the example (positive dictionary + code) since it looks like the components are a bit sparser than NMF.

Another idea is to check if it works as a topic model for text data, comparably to how NMF does.

ogrisel commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -1028,6 +1076,12 @@ class DictionaryLearning(BaseEstimator, SparseCodingMixin):
+        Whether to enforce positivity when finding the code.
+
+    fit_positive : bool
+        Whether to enforce positivity when finding the dictionary
```
Why not name this `code_positive` and `dict_positive` instead?

ogrisel commented on sklearn/decomposition/tests/test_dict_learning.py:
```diff
@@ -135,6 +145,25 @@ def test_dict_learning_online_shapes():
+    n_components = 8
+
+    dico = MiniBatchDictionaryLearning(
+        n_components, transform_algorithm='lasso_lars', random_state=0,
```
If the test is not too long to run, it would be great to also check with other values of `transform_algorithm`.

ogrisel commented on sklearn/decomposition/tests/test_dict_learning.py:
```diff
@@ -135,6 +145,25 @@ def test_dict_learning_online_shapes():
+                                            dict_positive=True,
+                                            code_positive=True)
+    assert_true((dictionary >= 0).all())
+    assert_true((code >= 0).all())
```
And maybe also test for other combinations such as: (`dict_positive=True` and `dict_positive=False`) and (`dict_positive=False` and `dict_positive=True`).

`

ogrisel reviewed (approved):
LGTM but disclaimer: I am not familiar with the SPAMS implementation. Maybe @arthurmensch is more knowledgeable.

It would be great to update the [faces decomposition example](http://scikit-learn.org/stable/auto_examples/decomposition/plot_faces_decomposition.html#sphx-glr-auto-examples-decomposition-plot-faces-decomposition-py) with a new entry with a positivity constraint on the dictionary components.

jakirkham commented:
Thanks for your feedback, @ogrisel.

Have pushed some changes that hopefully address your comments and have squashed the history a bit. Tests appear to be passing. 🎉

For the most part this is just relying on whatever algorithm exposed positivity constraint is available and exposing that to the user. With the exception of the thresholding case where it is custom and orthogonal matching pursuit where we just error out (as it is unsupported), which is just following @GaelVaroquaux's [advice earlier in the thread]( https://github.com/scikit-learn/scikit-learn/pull/6374#issuecomment-185064561 ). So would hope knowledge of SPAMS is not required, but it would be nice to have someone with that expertise take a look, if they have time.

As to an example (though we did look at this case in person), have the [faces with the positivity constraint in a notebook]( https://github.com/scikit-learn/scikit-learn/pull/6374#issuecomment-186991293 ). This shows a few different results depending on these constraints. Would be happy to add them if they would be useful. Could also play with different data like the digits set as you suggested or something else that we deem appropriate. Though would suggest either selecting one or two of these to add to the existing gallery or breaking them out on another doc page to avoid having too much info for users to ingest at one time.

Thoughts on any/all of this would be welcome. 😄

ogrisel commented:
Also, I am wondering if it would not be better to rename "dict_positive" to "positive_dict" and "code_positive" to "positive_code". You are the native English speaker though :)

jakirkham committed:
ENH: Add positivity option for code and dictionary

Provides an option for dictionary learning to positively constrain the
dictionary and the sparse code. This is useful in applications of
dictionary learning where the data is know to be positive (e.g. images),
but the sparsity constraint that dictionary learning has is better
suited for factorizing the data in contrast to other positively
constrained factorization techniques like NMF, which may not be
similarly sparse.

jakirkham committed:
TST: Test positivity with code and dictionary

Ensure that when the positivity constraint is applied that the
dictionary and code end up having only positive values in the respective
results depending on whether dictionary and/or code are positively
constrained.

jakirkham committed:
DOC: Positivity constraints dictionary learning

Shows the various positivity constraints on dictionary learning and what
the results of these look like using a Red to Blue color map. These are
included in the examples and also in the docs below dictionary learning.
All of these use the Olivetti faces as a training set.

jakirkham commented:
Thanks for the suggestions, @ogrisel.

Agree that `positive_dict`/`positive_code` sounds better. Have updated the API accordingly.

The red/blue color map with faces is a good idea. Have included some code for this in `plot_faces_decomposition.py` in a separate section. Also have included these images in the docs under the "Generic dictionary learning" section with some text beforehand.

Please let me know what you think. :)

jakirkham commented:
Have added a screenshot of what this doc section looks like below.

<details>

<img width="259" alt="Dictionary Learning Docs" src="https://user-images.githubusercontent.com/3019665/41377146-30cf201c-6f29-11e8-8f82-edda892f7983.png">

</details>

jakirkham commented:
Any other thoughts?

ogrisel commented:
Thank you very much @jakirkham! That is a great contribution.

ogrisel commented:
I merged a bit too quickly: I forgot to ask for an entry in whats_new.rst and we also need the `.. versionadded:: 0.20` annotation for the new options in the docstrings of the public functions / methods.

jakirkham commented:
Thanks @ogrisel. Glad to have this in. 😄

No worries. Does PR ( https://github.com/scikit-learn/scikit-learn/pull/11341 ) address this?

GaelVaroquaux commented:
Great! This is a fantastic addition. Thanks a lot!

jakirkham commented:
Of course. Thanks for the help reviewing.
